Load default configration from yaml file

vineetpalan opened the pull request.

amisstea commented on bin/review-rot:
```diff
@@ -1,57 +1,42 @@
     results = []
-    for item in config:
+
+    for item in config.get('git_services'):
```
If the user doesn't specify `git_services` in the config, this will result in a `NoneType` error. Perhaps defaulting to a list would be a nicer user experience.

```python
for item in config.get('git_services', []):
```

amisstea commented on bin/review-rot:
```diff
@@ -71,23 +56,134 @@ def main(state_, value, duration, ssl_verify=None):
+        # convert to new format
+        config = dict(git_services=config)
+        answer = str(raw_input("Would you like to rewrite '.reviewrot.yaml"
+                               "' in new format [Y/n] : ")).lower().strip()
```
This should probably be lowercase ("[y/n]") since there is no default value if the user only hits enter.

vineetpalan replied:
@amisstea What should be the ideal behavior if user hits enter? Should it rewrite?

amisstea replied:
You can probably default to rewrite.

vineetpalan replied:
Okay, got it

amisstea commented on bin/review-rot:
```diff
@@ -71,23 +56,134 @@ def main(state_, value, duration, ssl_verify=None):
+            log.info("Rewriting %r in new format!" % path_to_config)
+            with open(path_to_config, 'w') as f:
+                f.write(yaml.dump(config, default_flow_style=False))
+                f.write("\narguments:\n")
```
Instead of manually constructing yaml, why not just set `arguments` to `null`?

```python
config = dict(
    git_services=config,
    arguments=None
)
```

amisstea commented on bin/review-rot:
```diff
@@ -71,23 +56,134 @@ def main(state_, value, duration, ssl_verify=None):
+            with open(path_to_config, 'w') as f:
+                f.write(yaml.dump(config, default_flow_style=False))
+                f.write("\narguments:\n")
+                f.close()
```
No need to explicitly call `close`. The `open` context manager takes care of this.

amisstea commented on bin/review-rot:
```diff
@@ -71,23 +56,134 @@ def main(state_, value, duration, ssl_verify=None):
+             arguments (dict): Returns the parsed arguments
+     """
+    parsed_arguments = {}
+    command_line_args = vars(args)
```
Where is `args` defined? It needs to be passed into this function. Some unit tests would be helpful to detect this.

vineetpalan replied:
@amisstea  `args` is globally defined, so I was using it directly in method. Do I need to pass it into the function?

amisstea replied:
It would definitely be better to avoid the global variable if possible.

amisstea commented on bin/review-rot:
```diff
@@ -71,23 +56,134 @@ def main(state_, value, duration, ssl_verify=None):
+                    logged_error = True
+
+        if config_arguments.get('debug'):
+            log.setLevel(level=logging.DEBUG)
```
Why is this option treated differently than the rest? I would expect to be able to specify debug from the cli and have it override the value from the config. Also, to keep this function single purpose, it would be best to set a key/value pair in `parsed_arguments` rather than actually setting the log level here.

amisstea commented on bin/review-rot:
```diff
@@ -71,23 +56,134 @@ def main(state_, value, duration, ssl_verify=None):
+        if config_arguments.get('debug'):
+            log.setLevel(level=logging.DEBUG)
+
+        if config_arguments.get('reverse'):
```
Why can this not be specified via the cli?

vineetpalan replied:
This can be specified via cli
But if `--reverse` is specified from command line, the value will be true. If not, by default it will be false.
So if `reverse` is specified as `true` in config file, I am setting it manually

amisstea replied:
Ok this makes sense. Please add a comment explaining this.

vineetpalan replied:
Will do it

amisstea commented on bin/review-rot:
```diff
@@ -71,23 +56,134 @@ def main(state_, value, duration, ssl_verify=None):
+        if config_arguments.get('reverse'):
+            parsed_arguments['reverse'] = True
+
+        if parsed_arguments['ssl_verify'] is None:
```
Why can this not be specified via the cli?

vineetpalan replied:
From cli, you can specify `--insecure` flag or provide location of certificate using `--cacert`.
But from config file, it uses `ssl_verfiy` to specify the above details. This code was already present, I just moved it into the function.

amisstea replied:
I think it would be a good idea to align the cli and the config then.

vineetpalan replied:
Using the following code, we see what value is provided in cli
`Line 153: parsed_arguments['ssl_verify'] = False if args.insecure else args.cacert`
and if `parsed_arguments['ssl_verify'] is None`, then we go for config values.

amisstea replied:
What I meant was can we remove the `ssl_verify` option from the config and instead use both `insecure` and `cacert`?  You could also do the opposite and replace the combination of `insecure` and `cacert` with `ssl_verify` on the cli. I just don't see a reason to maintain two different sets of args for the same functionality.

amisstea commented on examples/sampleinput_new_format.yaml:
```diff
@@ -0,0 +1,34 @@
+      - project_name
+
+# Optional tag
+arguments: 
```
Missing ssl_verify option in this example.

ralphbean replied:
I see it down on line 32 now.

vineetpalan commented:
@amisstea @ralphbean @nirzari  I have added unit test and refactored the code. There were few bugs which I was able to identify during unit test. I have fixed those too. Please let me know if any test scenario needs to be added.

amisstea commented on reviewrot/__init__.py:
```diff
@@ -24,3 +35,156 @@ def get_git_service(git):
+            value (str): argument value
+            choices (dict): valid values of choices for arguments
+       Returns:
+             arguments (dict): Returns boolean value
```
I think you meant to put `arguments` under Args.

vineetpalan replied:
yea..my bad..will fix this

amisstea commented on reviewrot/__init__.py:
```diff
@@ -24,3 +35,156 @@ def get_git_service(git):
+
+    if config_arguments is not None:
+        for argument in config_arguments:
+                # Explicitly commandline arguments cannot be specified
```
This block has been indented a little too much.

vineetpalan replied:
Will fix this.

ralphbean commented on reviewrot/__init__.py:
```diff
@@ -24,3 +35,156 @@ def get_git_service(git):
+
+    # read input from the config file for pull requests
+    config = load_ordered_config(config_path)
+    if isinstance(config, list):
```
I'd like to see a test of this specific branch of code (if there isn't one already).

Something that calls `load_config_file` on an *old style* config, mocks out `raw_input` to say "yes", and then confirms that this function wrote a valid converted file back in place.

vineetpalan replied:
@ralphbean I have added 2 test cases related to config re-write

ralphbean commented on test/test.py:
```diff
@@ -190,5 +195,215 @@ def test_gerrit_request_reviews(self):
+                         and git_services_present)
+
+    @mock.patch('__builtin__.raw_input', return_value='y')
+    def test_load_config_file_re_write_yes(self, mocked_input):
```
Nice!
